torch/csrc/jit/codegen/cuda/transform_iter.cpp

```diff
@@ -440,6 +473,282 @@ int BestEffortReplay::findFirstMismatchedID(
   return td1->nDims();
 }
 
+namespace {
+
+// Maps that track information relevant to best effort replay about broadcast
+// axes in consumer that are not in producer
+//
+// For example if we have consumer: T0[i0, b1, b2, i3] and producer:
+// T1[i0, i3]
+//
+// If consumer transformations are:
+// -> T[i0, b1o, b1i, b2o, b2i, i3]
+// -> T[i0*b1i*b2o, b1o, b2i, i3]
+// -> T[i0*b1i*b2o*i3, b1o, b2i]
+//
+// forwarding_map would forward i0->i0*b1i*b2o
+// compliment_map would have the entry i0*b1i->b1o and
```

Comment:
I'm not sure why `compliment_map` would have those entries. IIUC, the map entry is created when a merge expression has one input that exist both in the producer and the consumer and another that only exist in the consumer. In such a case, a compliment map is created from the former input to the latter input. Here, `b1o` is not merged with any other domain, so I don't get why it shows up in the compliment map. For `i0*b1i`, I would expect it would be mapped to `b2o`, assuming there is a merge expression of `i0*b1i` and `b2o`.

Reply:
IIUC, `BestEffortReplayCasP`, `b1o` and `b2i` would be added to the leaf set.

Reply:
Yes, you're right. Will correct the comment. Thank you.

Reply:
Changed to:
```
// forwarding_map would forward i0->i0*b1i*b2o
// compliment_map would have the entry i0->b1o and i0*b1i->b2i
```
I think that should be right now.

Reply:
The forwarding map looks right, but souldn't the compliment map be:

```
i0->b1i
i0*b1i->b2o
```

`b1o` and `b2i` are not used in any of merge expressions, so they should not be `compliment_ids`.

Reply:
They are compliments of the forward, because if we forward `i0 -> i0*b1i` we would lose the leaf `b1o` because we never replayed the expression that takes `i0` and merges it with `b1i` which means we never replayed the split of `b1` going to `b1o` and `b1i`. Even though we never replayed that split in the best effort replay we want to account for the fact that `b1o` should be a leaf in the domain after the replay because of the forwarding.

Reply:
I understand that the intention here is to grab `b1o` and `b2i`, and I see how they are indeed grabbed eventually. My comment is just about `ConsumerForwardingInfo.compliment_map` and the above example case. `compliment_map` itself does not grab `b1o` and `b2i`. They are added to `leaf_ids` as outputs of `compliments_exprs` at around line 692.

I realize this is just a minor detail, but just wanted to make it less ambiguous.

Reply:
Ah okay, sorry for my confusion. This is pretty nuanced but I thought its usage to get the leaf ids in `replayCasP` was relatively clear. Do you have any suggestions to improve the comments? The principle is to just make sure what looked like a compliment wasn't used in another transformation expr which would mean it isn't a leaf.

Reply:
Yes, I understand what this class is designed to do now, but it was a little confusing as the comment wasn't exactly matching with my understanding of the logic, so I was unsure whether I was reading the code wrong or there was something wrong in the comment. 

I'd suggest to illustrate two things by the example: 1) the end result we need and 2) what this particular class does for that. In the example case, what we eventually want to do is to detect `b1o` and `b2i` as leaves. For that, `ConsumerForwardingInfo` creates mappings of `i0->b1i` and `i0*b1i->b2o`.